### PR TITLE
refactor(snowflake): port to sqlglot

### PR DESCRIPTION
**Pull request by cpcloud**
Port the snowflake backend over to sqlglot.

### CODE REVIEW

File: ibis/backends/snowflake/tests/test_client.py [hunk 1]
```diff
@@ -197,22 +190,6 @@ def test_read_csv_options(con, tmp_path):
     assert t.schema() == ibis.schema(dict(a="int64", b="int64"))
 
 
-def test_read_csv_https(con):
```
Comment (cpcloud) [inline ibis/backends/snowflake/tests/test_client.py:200]: Need to bring this test back.

Comment (cpcloud) [reply ibis/backends/snowflake/tests/test_client.py]: Done

**Comment by cpcloud** [comment]
Snowflake tests passing as of 93f757ccc1d24f36acb18fbcbc5f074ac9c35300:

```
…/ibis on  tes-snowflake is 📦 v7.2.0 via 🐍 v3.10.13 via ❄️  impure (ibis-3.10.13-env)
❯ pytest -m snowflake --snapshot-update -n 8 --dist loadgroup -q
bringing up nodes...
xxxxx.xx.x..xx.x.xx.x.xx.x...xx.x..s.x.....x...x...........x...xx.......x...............................................x.....................................x...............x.........x.x... [ 13%]
x.x...x.....xx..x.....x..xx..xx...x.x....xx..x....xx.....xx.xxx.x......x............x..................x...x.x.......x..x.........x...x.....................................x...x......x....x. [ 26%]
.....xx............x.....x...x...xx....x..............x.....xxxxx............x..x............xxx.........x.........x.x.....x.........x..........x.....x.x........xx.......x..........x..x.x... [ 39%]
x...x.x........x....x..x...x.x.xxxxx.xxx.xxxx.x...x...........x.....xx.....x......x...x...............x..x.......................x.....x................x.x.........x.........x.x.....x...x... [ 52%]
.x........x...x..x...x.x.x.........x.........x.......................x...x.............x..................x......................................x.....x..x...xx....s......................... [ 65%]
...x.....x.........x.........................x.......x.....x..........xx...xx.......x...x......xx................x......x.......x......x...............................x.....x................ [ 78%]
.............x.....x...................................x......x.............x..x..........x..............x.....x...xxx....xx.x....................x........................................... [ 91%]
.............s................................................................................................................                                                                 [100%]
1258 passed, 3 skipped, 195 xfailed in 136.59s (0:02:16)
```

File: ibis/backends/oracle/__init__.py [hunk 2]
```diff
@@ -16,11 +16,10 @@
...
 # Get out your clippers, it's time to shave a yak.
 #
-# 1. snowflake-sqlalchemy doesn't support sqlalchemy 2.0
-# 2. oracledb is only supported in sqlalchemy 2.0
-# 3. Ergo, module hacking is required to avoid doing a silly amount of work
+# 1. oracledb is only supported in sqlalchemy 2.0
+# 2. Ergo, module hacking is required to avoid doing a silly amount of work
 #    to create multiple lockfiles or port snowflake away from sqlalchemy
-# 4. Also the version needs to be spoofed to be >= 7 or else the cx_Oracle
+# 3. Also the version needs to be spoofed to be >= 7 or else the cx_Oracle
```
Comment (cpcloud) [inline ibis/backends/oracle/__init__.py:22]: Once we port the remaining SQL backends I believe we can remove this hack.

**Comment by cpcloud** [comment]
Snowflake tests looking good after a rebase:

```
❯ pytest -m snowflake --snapshot-update -n 8 --dist loadgroup -q
bringing up nodes...
.x...x............x.x...x..x..........x.x..x.xxx.....x......x.x...s..x.x......x......x.........x.............x........................x......x........x.............x......x..x........... [ 12%]
.....x.....x....x......x......x.xxx...xx...................x.x....xx.....x....x....x...x.....x.x..........x.x.......x.x.x........................x..xx..x.x......x.....x...x...x.....x..xx [ 25%]
.x..x.......x.........x.x.x................x..................x....x.....x..................xx..........x....x....x..x.x............x...x....xx.x.x...........xxx.x..........x..xx..x.x... [ 38%]
.x.......x...x...............................x...........x........x.x....................xx...xx......x..x.......x..x.x.x....x..x.x..x......................x........x.xx....x......x.x..x [ 51%]
.....x......................x..x.......xx.............x................................x..........x..................x.................x...........................xx......x...x..x....... [ 63%]
.............x............x.............x....x.xxx...x..x.xxxxxxx......x.....x.....x...x..x.x.....x..x..x...........x..x..............s.......x...........................x...x........... [ 76%]
......x...x..x...........xx...x......................x...........x........x...x.....x...........xx.x..x..........x.....x.....x.xx...........xx.....x.....x.....x.........................s [ 89%]
...........................................................................................................................................................                                [100%]
1259 passed, 3 skipped, 195 xfailed in 131.72s (0:02:11)
```

File: ibis/backends/base/sqlglot/rewrites.py [hunk 3]
```diff
@@ -104,7 +114,7 @@ def merge_select_select(_):
         if v.find((Window, ops.Unnest), filter=ops.Value):
             return _
     for v in _.predicates:
-        if v.find(ops.ExistsSubquery, filter=ops.Value):
+        if v.find(ops.Subquery, filter=ops.Value):
```
Comment (kszucs) [inline ibis/backends/base/sqlglot/rewrites.py:117]: What is the reason of this change?

File: ibis/backends/tests/sql/snapshots/test_compiler/test_agg_filter/out.sql [hunk 4]
```diff
@@ -1,10 +1,16 @@
 SELECT
-  t0.a,
-  t0.b * CAST(2 AS TINYINT) AS b2
-FROM my_table AS t0
+  t1.a,
+  t1.b2
+FROM (
+  SELECT
```
Comment (kszucs) [inline ibis/backends/tests/sql/snapshots/test_compiler/test_agg_filter/out.sql:5]: These would nice to be squashed, is this change relevant to the rewrite change?

File: ibis/backends/base/sqlglot/rewrites.py (hunk 3 above)
Comment (cpcloud) [reply ibis/backends/base/sqlglot/rewrites.py:117]: `InSubquery` has the same limitations as `ExistsSubquery` and without this change the generated code for tpch query 20 is incorrect.

This seems be an issue when using a projection for the leaf nodes (see `_tpch_table` for the Snowflake backend)

File: ibis/backends/tests/sql/snapshots/test_compiler/test_agg_filter/out.sql (hunk 4 above)
Comment (cpcloud) [reply ibis/backends/tests/sql/snapshots/test_compiler/test_agg_filter/out.sql:5]: Yep, I believe it is.

**Comment by cpcloud** [comment]
Snowflake tests are passing:

```
…/ibis on  tes-snowflake is 📦 v7.2.0 via 🐍 v3.10.13 via ❄️  impure (ibis-3.10.13-env) took 8s
❯ pytest -m snowflake --snapshot-update -n 8 --dist loadgroup -q
bringing up nodes...
......xx...x.x.x...x...x...xx.......x....x..x..x.......x..xxx.x........x.x...x.x.x.x..x.xx.....x..x...............x........xx....x..x......x.....x.....x.........x...x..........x.x........... [ 13%]
.....x........x............x..x.....x.......x.....x.....x..x.....x............x.....................................x.x.........x..x....................s....x...x....x.xx.x..xxxxxxxx.......x [ 26%]
.x....x......x........................x.xx.x...x..........x.x....x.x.......x..............x........x............x....x......xx..........x.....x...........x.........................x......... [ 39%]
.......x.x.............x.........x...xx..x.x....x.................x........x...x........xx..........x.........x.........................x.....xxxx...x.x..........x.....x.............x.....x. [ 52%]
.xxx....x.....x.x....................x.x...x.................xx.........x..x.......xx..x..x.............x..x.........xx.x.......x.......xx.....xxx....x..x...................xx............... [ 65%]
.x................................x.....x..............x......x...............x........x.x............xx....x..xx.x.....x..........x.x............x.....x...x.x.............x.....x.......x.x. [ 78%]
.........x....x..x....x.......xx........x........xx...x....xxx........................................x....x............s......x..x........................................................... [ 91%]
......................................s........................................................................................                                                                [100%]
1259 passed, 3 skipped, 195 xfailed in 157.35s (0:02:37)
```

**Comment by cpcloud** [comment]
Snowflake tests passing after a rebase on latest `upstream/the-epic-split`:

```
…/ibis on  tes-snowflake is 📦 v7.2.0 via 🐍 v3.10.13 via ❄️  impure (ibis-3.10.13-env)
❯ pytest -m snowflake --snapshot-update -n 8 --dist loadgroup -q
bringing up nodes...
...............x...x...............x...............x....x.....................x....x......x....x.x.......x.....x...x....xx.x....x.x........x...x....x......x.x....x...x......x................ [ 13%]
......x.....x............x...x........x.........x.....x.....x.x......x.........x..xx.....x....x.x....xx..x...........x..xx......x.x.....x.........xx..xx.x.....x..x...xxxx..xx..x....x.x...... [ 26%]
x.....x.x....................x..........x.xx.xx...x...x.xx....xx.x.....x..x.x.xx...xx......xx..xx....xx.x....xx....x....x.........x...............x........x.................................x [ 39%]
.................xx................x...x...........................................................x......................................xx.....x.x................x....xx...............x... [ 52%]
...x........x......................................xx..x......x............s.........x..x...........x.......x..........x...x..............s.........xxxxxx.xx...x...xxx.x.......x......x...... [ 65%]
.........x.........x....x..................x..........x.......x........x......xx........x.xxx.x.............x...x.................x..x......x...xxx...xx........x........xxx......x...x....... [ 78%]
.......x....x...............x...x......x.......x....x...x.xxx.......x...........x.....x....................x..x..x....x.......x..............x................................................ [ 91%]
..............................................................................................................s................                                                                [100%]
1259 passed, 3 skipped, 195 xfailed in 129.48s (0:02:09)
```

File: ibis/backends/snowflake/__init__.py [hunk 5]
```diff
@@ -947,43 +999,46 @@ def read_parquet(
...
-    https://docs.snowflake.com/en/release-notes/bcr-bundles/2023_04/bcr-1057
-    even if they are trivial boolean literals.
-    """
-    result = compiler.visit_join(element, **kw)
 
-    if element.right._is_lateral:
-        return re.sub(r"^(.+) ON true$", r"\1", result, flags=re.IGNORECASE | re.DOTALL)
-    return result
+# @compiles(sa.sql.Join, "snowflake")
+# def compile_join(element, compiler, **kw):
```
Comment (kszucs) [inline ibis/backends/snowflake/__init__.py:1033]: Should this be removed?

Comment (cpcloud) [reply ibis/backends/snowflake/__init__.py:1033]: Yep, thanks!

Comment (cpcloud) [reply ibis/backends/snowflake/__init__.py]: Done.

**Review by kszucs — APPROVED** [review]
The compiler looks way more straightforward now, nice!
